File: app/src/main/java/me/devsaki/hentoid/util/file/PermissionHelper.java
package me.devsaki.hentoid.util.file;

import static android.content.pm.PackageManager.PERMISSION_GRANTED;

import android.Manifest;
import android.app.Activity;
import android.os.Build;

import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

public class PermissionHelper {

    public static final int RQST_STORAGE_PERMISSION = 3;
    public static final int RQST_NOTIFICATION_PERMISSION = 4;


    private PermissionHelper() {
        throw new IllegalStateException("Utility class");
    }

    public static boolean requestExternalStorageReadPermission(Activity activity, int permissionRequestCode) {
        if (ContextCompat.checkSelfPermission(activity,
                Manifest.permission.READ_EXTERNAL_STORAGE) == PERMISSION_GRANTED) {
            return true;
        } else {
            ActivityCompat.requestPermissions(activity, new String[]{
                    Manifest.permission.READ_EXTERNAL_STORAGE}, permissionRequestCode);
            return false;
        }
    }

    public static boolean checkExternalStorageReadWritePermission(Activity activity) {
        return (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU
                || (ContextCompat.checkSelfPermission(activity,
                Manifest.permission.READ_EXTERNAL_STORAGE) == PERMISSION_GRANTED
                && ContextCompat.checkSelfPermission(activity,
                Manifest.permission.WRITE_EXTERNAL_STORAGE) == PERMISSION_GRANTED)
        );
    }

    public static boolean requestExternalStorageReadWritePermission(Activity activity, int permissionRequestCode) {
        if (checkExternalStorageReadWritePermission(activity) || Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU)
            return true;
        else {
            ActivityCompat.requestPermissions(activity, new String[]{
                    Manifest.permission.READ_EXTERNAL_STORAGE, Manifest.permission.WRITE_EXTERNAL_STORAGE}, permissionRequestCode);
            return false;
        }
    }

    public static boolean checkNotificationPermission(Activity activity) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            return ContextCompat.checkSelfPermission(activity, Manifest.permission.POST_NOTIFICATIONS) == PERMISSION_GRANTED;
        } else return true;
    }

    public static boolean requestNotificationPermission(Activity activity, int permissionRequestCode) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            if (checkNotificationPermission(activity)) return true;
            else {
                ActivityCompat.requestPermissions(activity, new String[]{Manifest.permission.POST_NOTIFICATIONS}, permissionRequestCode);
                return false;
            }
        } else return true;
    }
}

File: app/src/main/java/me/devsaki/hentoid/util/file/PermissionHelper.kt
package me.devsaki.hentoid.util.file

import android.Manifest.permission.POST_NOTIFICATIONS
import android.Manifest.permission.READ_EXTERNAL_STORAGE
import android.Manifest.permission.WRITE_EXTERNAL_STORAGE
import android.app.Activity
import android.content.Context
import android.content.pm.PackageManager.PERMISSION_GRANTED
import android.os.Build
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat

const val RQST_STORAGE_PERMISSION = 3
const val RQST_NOTIFICATION_PERMISSION = 4

private fun Context.checkPermission(code: String): Boolean {
    return ContextCompat.checkSelfPermission(this, code) == PERMISSION_GRANTED
}

fun Activity.requestExternalStorageReadPermission(permissionRequestCode: Int): Boolean {
    return if (checkPermission(READ_EXTERNAL_STORAGE)) true
    else {
        ActivityCompat.requestPermissions(
            this, arrayOf(READ_EXTERNAL_STORAGE),
            permissionRequestCode
        )
        false
    }
}

fun Activity.checkExternalStorageReadWritePermission(): Boolean {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) return true
    return checkPermission(READ_EXTERNAL_STORAGE) &&
            checkPermission(WRITE_EXTERNAL_STORAGE)
}

fun Activity.requestExternalStorageReadWritePermission(permissionRequestCode: Int): Boolean {
    return if (checkExternalStorageReadWritePermission() || Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) true else {
        ActivityCompat.requestPermissions(
            this, arrayOf(
                READ_EXTERNAL_STORAGE,
                WRITE_EXTERNAL_STORAGE
            ), permissionRequestCode
        )
        false
    }
}

fun Context.checkNotificationPermission(): Boolean {
    return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
        checkPermission(POST_NOTIFICATIONS)
    } else true
}

fun Activity.requestNotificationPermission(permissionRequestCode: Int): Boolean {
    return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
        if (checkNotificationPermission()) true else {
            ActivityCompat.requestPermissions(
                this,
                arrayOf(POST_NOTIFICATIONS),
                permissionRequestCode
            )
            false
        }
    } else true
}